Support skip invalid journal record in replying journal stage

hangc0276 opened the pull request:
### Motivation
This PR is generated from https://github.com/apache/bookkeeper/pull/3437
Fix https://github.com/apache/bookkeeper/issues/2220

When we use `kill -9` command to stop one bookie pod, the journal file may be broken. If we try to start up the bookie pod again, the pod will startup to fail with the following exception.
```
10:15:55.026 [main] ERROR org.apache.bookkeeper.bookie.Bookie - Exception while replaying journals, shutting down
java.io.IOException: Invalid record found with negative length -448299468
        at org.apache.bookkeeper.bookie.Journal.scanJournal(Journal.java:821) ~[org.apache.bookkeeper-bookkeeper-server-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.bookie.Bookie.replay(Bookie.java:945) ~[org.apache.bookkeeper-bookkeeper-server-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.bookie.Bookie.readJournal(Bookie.java:911) ~[org.apache.bookkeeper-bookkeeper-server-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.bookie.Bookie.start(Bookie.java:965) ~[org.apache.bookkeeper-bookkeeper-server-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.proto.BookieServer.start(BookieServer.java:156) ~[org.apache.bookkeeper-bookkeeper-server-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.server.service.BookieService.doStart(BookieService.java:68) ~[org.apache.bookkeeper-bookkeeper-server-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.common.component.AbstractLifecycleComponent.start(AbstractLifecycleComponent.java:83) ~[org.apache.bookkeeper-bookkeeper-common-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.common.component.LifecycleComponentStack.lambda$start$4(LifecycleComponentStack.java:144) ~[org.apache.bookkeeper-bookkeeper-common-4.12.0.jar:4.12.0]
        at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:406) [com.google.guava-guava-30.1-jre.jar:?]
        at org.apache.bookkeeper.common.component.LifecycleComponentStack.start(LifecycleComponentStack.java:144) [org.apache.bookkeeper-bookkeeper-common-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.common.component.ComponentStarter.startComponent(ComponentStarter.java:85) [org.apache.bookkeeper-bookkeeper-common-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.server.Main.doMain(Main.java:234) [org.apache.bookkeeper-bookkeeper-server-4.12.0.jar:4.12.0]
        at org.apache.bookkeeper.server.Main.main(Main.java:208) [org.apache.bookkeeper-bookkeeper-server-4.12.0.jar:4.12.0]
10:15:55.134 [main] INFO  org.apache.zookeeper.ZooKeeper - Session: 0x200064a14681be2 closed
```

We should have a way to make the bookie pod startup instead of decommissioning it.


### Changes
1. Add a configuration to allow skipping the invalid entries when journal replying to journal files.

hangc0276 commented:
@frankjkelly I have created a new PR to fix the bookie that can't startup due to the journal reply failing.

eolivelli commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java:
```diff
@@ -4010,6 +4012,25 @@ public boolean isDataIntegrityStampMissingCookiesEnabled() {
+     * @see #isSkipReplayJournalInvalidRecord .
+     */
+    public boolean isSkipReplayJournalInvalidRecord() {
+        return this.getBoolean(SKIP_REPLAY_JOURNAL_INVALID_RECORD, false);
```
the previous behaviour was to skip errors, we should keep the same value

hangc0276 replied:
The previous behavior is to throw the exception out and block the bookie start-up. We should keep the same behavior.

hangc0276 replied:
Even though the previous default behavior is `false`, I changed the default value to `true`, because I think it's valuable to skip the the invalid record instead of make the bookie node can't start up.

eolivelli reviewed (changes requested):
Overall LGTM, but I would set the property to true by default

wenbingshen commented on bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java:
```diff
@@ -827,22 +828,32 @@ public long scanJournal(long journalId, long journalPos, JournalScanner scanner)
-                        fullRead(recLog, lenBuff);
-                        if (lenBuff.remaining() != 0) {
-                            break;
+                    try {
+                        if (len == PADDING_MASK && journalVersion >= JournalChannel.V5) {
+                            // skip padding bytes
+                            lenBuff.clear();
+                            fullRead(recLog, lenBuff);
+                            if (lenBuff.remaining() != 0) {
+                                break;
+                            }
+                            lenBuff.flip();
```
line 820 may throw IOException, this may cause bookie startup failed ==>  fullRead(recLog, lenBuff);
In this PR, line 835 also may throw IOException, because this future the bookie can continue starting.
Same behavior here, two results, right?

https://github.com/apache/bookkeeper/pull/3956/files#diff-0a79d54e3dfd4bab1d85510ef957f8f4b7ab6f6acd09aadd634b4d7bbee6e1f4R815-R821
```java
            while (true) {
                // entry start offset
                long offset = recLog.fc.position();
                // start reading entry
                lenBuff.clear();
                fullRead(recLog, lenBuff);
```

hangc0276 replied:
I caught all the `fullRead` throw exceptions.

wenbingshen reviewed (approved):
LGTM

wenbingshen commented:
rerun failure checks

frankjkelly commented:
Hi folks - thanks for the efforts here - what are the remaining issues on this? Thanks!

zymap commented:
@eolivelli Would you like to take another look at this PR?

hangc0276 commented:
> Hi folks - thanks for the efforts here - what are the remaining issues on this? Thanks!

@frankjkelly I addressed all the comments, please help take a look, thanks.

eolivelli reviewed (approved):
+1

vitalii-buchyn-exa commented:
hello community,
any documentation update for this feature?
